### ISSUE TITLE
Add Sample_weights to setup

### ISSUE THREAD

**Issue by jackmau** [pycaret/pycaret#1350]
Following my reply on LinkedIn to @moezali1 call for new features, here is a big one

**Is your feature request related to a problem? Please describe.**
Since #573 it has been possible to fit models with `sample_weights` argument. However sample_weights are consequently ignored in pycaret metrics, evaluation and are not used to split the data at `setup` stage.

**Describe the solution you'd like**
I propose, as originally suggested in https://github.com/pycaret/pycaret/issues/573#issuecomment-684041587, to implement a `sample_weight` argument in the `setup` function which would be _automagically_ used:
- to split the data into train/validation/test (currently NOT in scikit learn, see https://github.com/scikit-learn/scikit-learn/issues/4632#issuecomment-393945555 for possible implementation)
- as an argument to fit, possibly removing all the scikit learn regressors/classifiers that do not support `sample_weights` by default in `compare_models`
- as an argument to all metrics, note that, for classification problem AUC doesn't have a sample weight problems, so you'll need a whole new AUC definiton, similar to the one in [this example in the scikit-learn documentation](https://scikit-learn.org/stable/auto_examples/linear_model/plot_poisson_regression_non_normal_loss.html#evaluation-of-the-ranking-power)
- in the model evaluation graphs (this probably at a later stage and in a future issue, since is much less urgent that the others)

**Describe alternatives you've considered**
I did try to create my own metrics with `add_metric` and using train data from `get_config`, but I messed it up badly (I think beacuse of inteornal cross val splits). I don't see proper `sample_weights` implementation as something doable _on top of_ pycaret without proper implementation _in_ pycaret.

**Additional context**
`Sample_weights` argument is key in most insurance/credit risk modeling problems, known under the name of _exposure_ particularly when modeling frequency of losses, which typically is representable by a poisson process. The assumption that underlies the poisson process is that the probabilty to observe an event is proportional to the time of the observation. Length of the contract, i.e. the previously mentioned _exposure_ is thus fundamental when modeling. A good example of practical use of `sample_weight` (with metrics using `sample_weight`, but with no validation split) is available in the [scikit learn documentation](https://scikit-learn.org/stable/auto_examples/linear_model/plot_poisson_regression_non_normal_loss.html#sphx-glr-auto-examples-linear-model-plot-poisson-regression-non-normal-loss-py)

**Comment by Prashavu** [comment]
@jackmau how do we give `sample_weights` i in `fit`or classification problems in the 3.0.0rc3 version of pycaret?

**Comment by tevfikcagridural** [comment]
Based on @mydataag’s answer it is possible give sample weights not on setup but on create_model or compare_models within pycaret. Works for both classification and regression

```
import numpy as np
from pycaret.classification import setup, compare_models, get_config 
from pycaret.datasets import get_data 

# Load dataset 
df = get_data('iris')

# Initialize pycaret training environment
setup(df, target='species')

# Weights as pandas Series for 
# all samples based on the df given in setup 
weights = np.random.rand(df.shape[0]) 

# Get index of train data  
idx = get_config('X_train').index

# Compare models with subset set in setup 
compare_models(fit_kwargs={'sample_weight': weights[idx]})
```

**Comment by alvarado1984** [comment]
If add class for example fit_kwargs={'class_weight': {0: 0.6118484712814661, 1: 2.735166892633482}}) the result is []  , Its not accept this variable

**Comment by alvarado1984** [comment]
In this case compare_models(fit_kwargs={'class_weight': dict(zip(df['species'].unique(), [1,2,3]))} the result is []

**Comment by ayansengupta17** [comment]
I faced the same issue as mentioned by @alvarado1984

**Comment by adrinjalali** [comment]
This might be of interest: https://github.com/scikit-learn/scikit-learn/pull/26896

**Comment by tiefenthaler** [comment]
Any updates related to this topic?

**Comment by jmakov** [comment]
> In this case compare_models(fit_kwargs={'class_weight': dict(zip(df['species'].unique(), [1,2,3]))} the result is []
Also my experience with 3.3.2.

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
